src/NUnitEngine/nunit.engine.core/Internal/TestAssemblyResolver.cs

```diff
@@ -2,171 +2,251 @@
+        {
+            // First, looking only at direct references by the test assembly, try to determine if
+            // this assembly is using WindowsDesktop (either SWF or WPF) and/or AspNetCore.
+            AssemblyDefinition assemblyDef = AssemblyDefinition.ReadAssembly(testAssemblyPath);
+            bool isWindowsDesktop = false;
+            bool isAspNetCore = false;
+            foreach (var reference in assemblyDef.MainModule.GetTypeReferences())
+            {
+                string fn = reference.FullName;
+                if (fn.StartsWith("System.Windows.") || fn.StartsWith("PresentationFramework"))
+                    isWindowsDesktop = true;
+                if (fn.StartsWith("Microsoft.AspNetCore."))
+                    isAspNetCore = true;
+            }
+
+            // Initialize the list of ResolutionStrategies in the best order depending on
+            // what we learned.
+            ResolutionStrategies = new List<ResolutionStrategy>();
+            
+            if (isWindowsDesktop && Directory.Exists(WINDOWS_DESKTOP_DIR))
+                ResolutionStrategies.Add(new AdditionalDirectoryStrategy(WINDOWS_DESKTOP_DIR));
+            if (isAspNetCore && Directory.Exists(ASP_NET_CORE_DIR))
+                ResolutionStrategies.Add(new AdditionalDirectoryStrategy(ASP_NET_CORE_DIR));
+            ResolutionStrategies.Add(new TrustedPlatformAssembliesStrategy());
+            ResolutionStrategies.Add(new RuntimeLibrariesStrategy(loadContext, testAssemblyPath));
+            if (!isWindowsDesktop && Directory.Exists(WINDOWS_DESKTOP_DIR))
+                ResolutionStrategies.Add(new AdditionalDirectoryStrategy(WINDOWS_DESKTOP_DIR));
+            if (!isAspNetCore && Directory.Exists(ASP_NET_CORE_DIR))
+                ResolutionStrategies.Add(new AdditionalDirectoryStrategy(ASP_NET_CORE_DIR));
+        }
+
         public void Dispose()
         {
             _loadContext.Resolving -= OnResolving;
         }
 
-        public Assembly Resolve(AssemblyLoadContext context, AssemblyName name)
+        public Assembly Resolve(AssemblyLoadContext context, AssemblyName assemblyName)
+        {
+            return OnResolving(context, assemblyName);
+        }
+
+        private Assembly OnResolving(AssemblyLoadContext loadContext, AssemblyName assemblyName)
         {
-            return OnResolving(context, name);
+            if (loadContext == null) throw new ArgumentNullException("context");
+
+            Assembly loadedAssembly;
+            foreach (var strategy in ResolutionStrategies)
+                if (strategy.TryToResolve(loadContext, assemblyName, out loadedAssembly))
+                    return loadedAssembly;
+
+            log.Info("Cannot resolve assembly '{0}'", assemblyName);
+            return null;
         }
 
-        private Assembly OnResolving(AssemblyLoadContext context, AssemblyName name)
+        #region Nested ResolutionStrategy Classes
+
+        public abstract class ResolutionStrategy
```

Comment:
I started out with a common constuctor, since removed, and I imagined there would be some shared implementation as well. Turns out that the approaches have very little in common right now but I decided to keep the class because I plan to add collection of stats there.

From debugging, I know that the RuntimeLibraries approach originally found most things but the addition of the trusted platforms section changed that and now that finds most things but not the Windows, WPF or AspNetCore code. It would be great if we could eventually understand this better and simplify it.